drivers/usb/udc/udc_renesas_ra.c

```diff
@@ -0,0 +1,754 @@
+/*
+ * Copyright (c) 2024 Renesas Electronics Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "udc_common.h"
+
+#include <soc.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/clock_control/renesas_ra_cgc.h>
+#include <zephyr/drivers/usb/udc.h>
+#include "r_usb_device.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(udc_renesas_ra, CONFIG_UDC_DRIVER_LOG_LEVEL);
+
+struct udc_renesas_ra_config {
+	const struct pinctrl_dev_config *pcfg;
+	size_t num_of_eps;
+	struct udc_ep_config *ep_cfg_in;
+	struct udc_ep_config *ep_cfg_out;
+	void (*make_thread)(const struct device *dev);
+	int speed_idx;
+};
+
+struct udc_renesas_ra_data {
+	struct k_thread thread_data;
+	struct st_usbd_instance_ctrl udc;
+	struct st_usbd_cfg udc_cfg;
+};
+
+enum udc_renesas_ra_event_type {
+	/* An event generated by the HAL driver */
+	UDC_RENESAS_RA_EVT_HAL,
+	/* Shim driver event to trigger next transfer */
+	UDC_RENESAS_RA_EVT_XFER,
+	/* Let controller perform status stage */
+	UDC_RENESAS_RA_EVT_STATUS,
+};
+
+struct udc_renesas_ra_evt {
+	enum udc_renesas_ra_event_type type;
+	usbd_event_t hal_evt;
+	uint8_t ep;
+};
+
+K_MSGQ_DEFINE(drv_msgq, sizeof(struct udc_renesas_ra_evt), CONFIG_UDC_RENESAS_RA_MAX_QMESSAGES,
+	      sizeof(uint32_t));
+
+extern void usb_device_isr(void);
+
+static void udc_event_handler(usbd_callback_arg_t *p_args)
+{
+	struct udc_renesas_ra_evt evt = {
+		.type = UDC_RENESAS_RA_EVT_HAL,
+		.hal_evt = p_args->event,
+	};
+
+	k_msgq_put(&drv_msgq, &evt, K_NO_WAIT);
+}
+
+void usb_interrupt_handler(void *arg)
+{
+	ARG_UNUSED(arg);
+	usb_device_isr();
```

Comment:
The `usb_device_isr` is implemented in the Renesas HAL module as a low-level driven routine with direct access to registers. Other processing for UDC functionality is still handled in the shim driver, based on hardware events that this ISR provides via callbacks. It seems our implementation is quite similar to others that have external modules for the HAL layer, such as NXP in `drivers/serial/uart_nxp_s32_linflexd.c` or STM in `drivers/usb/udc/udc_stm32.c`.

Reply:
> It seems our implementation is quite similar to others that have external modules for the HAL layer, such as NXP in `drivers/serial/uart_nxp_s32_linflexd.c` or STM in `drivers/usb/udc/udc_stm32.c`.

No your implementation is not similar to `drivers/usb/udc/udc_stm32.c`. I do not understand why you are referring to the NXP UART driver. Your shim driver design does not look good, and you will probably have problems with controllers that support high speed.

Reply:
I mean I see this as a common way when a shim driver, not only USB driver but also other drivers, uses the HAL API to implement its functionality. In that case, the HAL ISR just handle the low-level hardware routines when an interrupt occurs, get the interrupt event and notify the upper layer by callback.

> and you will probably have problems with controllers that support high speed.

For this comment, could you give me more details about your concern? Did you mention that it might cause this driver to have low performance and not be able to adapt to high-speed USB? Actually, I have tested this driver can work well as a high-speed device with currently supported samples (cdc_acm, mass, hid, ...)

Reply:
For example https://github.com/zephyrproject-rtos/zephyr/pull/79989#discussion_r1805420830

Reply:
> For example [#79989 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/79989#discussion_r1805420830)

I agreed with this comment and already have an update for it in local. I mean what about this topic? I think we are not the first to implement low level ISR somewhere other than shim driver, but in HAL module.

Reply:
Please push your changes.